Add `test-utils` module

zen-dog opened the pull request:
Summary:
- ported Marathon integration test classes. Most notably `MesosTest`, `ZookeeperServer` and `MesosFacade` and their dependencies
- ported `MesosClientIntegrationTest` into the `mesos-client` project. It depends on the above-mentioned test classes and starts Mesos/Zookeeper to run its tests.

JIRA: DCOS-47483

zen-dog committed:
Added `test-utils` module

Summary:
- ported Marathon integration test classes. Most notably `MesosTest`, `ZookeeperServer` and `MesosFacade` and their dependencies
- ported `MesosClientIntegrationTest` into the `mesos-client` project. It depends on the above mentioned test classes and starts Mesos/Zookeeper to run it's tests.

JIRA: DCOS-47483

zen-dog commented:
Most of the classes here are copy-pasted from Marathon with minimal changes. I removed some unused parts and cleaned up obvious low-hanging fruits. This PR depends on #24 (for provisioning Mesos and Zookeeper)

zen-dog committed:
Testing a different way to look up `detect-routable-ip.sh`

kensipe reviewed (dismissed):
Blocker:  Doesn't work with Java 11

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by javassist.ClassPool (file:/Users/kensipe/.gradle/caches/modules-2/files-2.1/org.javassist/javassist/3.18.1-GA/d9a09f7732226af26bf99f19e2cffe0ae219db5b/javassist-3.18.1-GA.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int)
WARNING: Please consider reporting this to the maintainers of javassist.ClassPool
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

alenkacz commented on build.gradle:
```diff
@@ -37,20 +37,29 @@ allprojects {
         compile group: 'com.typesafe.akka', name: "akka-http_$scalaVersion", version: '10.0.11'
+
+        // Apache Curator. See this [compatibility documentation](http://curator.apache.org/zk-compatibility.html)
+        def withoutZk = { exclude group: 'org.apache.zookeeper', module: 'zookeeper' }
```
should really all projects depend on curator? Even metrics? I think we should move this to `test-utils/build.gradle` right now

zen-dog replied:
I did that initially but then I realized that `test-utils`, `persistence-zookeeper`, frameworks and future leader election code will depend on it so I decided to move it up. wdyt?

alenkacz commented on test-utils/src/main/scala/com/mesosphere/utils/http/AkkaHttpResponse.scala:
```diff
@@ -0,0 +1,37 @@
+import scala.concurrent.{ExecutionContext, Future}
+import scala.reflect.ClassTag
+
+object AkkaHttpResponse {
```
do we need this to test library? There might not even be akka http, right?

alenkacz replied:
oh I see, we use this to talk to mesos API, that was not really obvious at first sight to me :)

jeschkies commented on build.gradle:
```diff
@@ -37,20 +37,29 @@ allprojects {
         // Use Scalatest for testing our library
         testImplementation 'junit:junit:4.12'
-        testImplementation "org.scalatest:scalatest_$scalaVersion:3.0.5"
+        compile "org.scalatest:scalatest_$scalaVersion:3.0.5"
```
Isn't this `testCompile`?

zen-dog replied:
`MesosTest` uses `eventually` ¯\_(ツ)_/¯

jeschkies replied:
Sure but why make it a global dependency? Just make it a local compile dependency for `test-utils`.

zen-dog replied:
Ah, got ya. Done

alenkacz replied:
good catch :)

jeschkies commented on mesos-client/src/test/scala/com/mesosphere/mesos/client/MesosClientIntegrationTest.scala:
```diff
@@ -0,0 +1,147 @@
+  with Eventually {
+
+  "Mesos client should successfully subscribe to mesos without framework Id" in withFixture() { f =>
+    When("a framework subscribes")
```
Where is the call?

zen-dog replied:
Implicitly in `withFixture` but I'll remove it anyway

jeschkies commented on test-utils/src/main/resources/detect-routable-ip.sh:
```diff
@@ -0,0 +1,16 @@
+# Script which returns your primary IP; works on Mac OS X and Linux
+#
+
+case $(uname) in
```
@timcharper what is the reason for this being as script and not Scala code?

timcharper replied:
The motivation at the time I wrote this was I could write this script and just run it on different distros in docker and check if it worked :)

If we can write equivalent Scala code then no problem there.

jeschkies replied:
Hm, `java.net.InetAddress.getLocalHost.getHostAddress()` returns the same as `detec-routable-ip.sh`. At least on my machine.

zen-dog replied:
Well, if it works on your machine, then we're good to go 😉

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/ProcessOutputToLogStream.scala:
```diff
@@ -0,0 +1,12 @@
+
+import scala.sys.process.ProcessLogger
+
+case class ProcessOutputToLogStream(process: String) extends ProcessLogger with StrictLogging {
```
Could you add some docs?

zen-dog replied:
Hu? To some internal 4-liner test utils class?

jeschkies replied:
Yep. This class is not trivial and it would help to understand.

zen-dog replied:
I'm open to `suggestion`s

jeschkies commented on mesos-client/src/test/scala/com/mesosphere/mesos/client/MesosClientIntegrationTest.scala:
```diff
@@ -0,0 +1,147 @@
+@RunWith(classOf[JUnitRunner])
+class MesosClientIntegrationTest extends AkkaUnitTest
+  with MesosClusterTest
+  with Eventually {
```
`Eventually` is already part of `AkkaUnitTest`.

alenkacz reviewed (approved):
LGTM :) I did not really read it line by line and some parts definitely look like the code could be improved but since it's battle tested and working I am fine with this :)

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/UnitTestLike.scala:
```diff
@@ -0,0 +1,76 @@
+  with Eventually
+  with TimeLimitedTests {
+
+  private class LoggingInformer(info: Informer) extends Informer {
```
```suggestion
  /**
   * Prepends logs with a position. See [[UnitTestLike.info]].
   */
  private class LoggingInformer(info: Informer) extends Informer {
```

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/UnitTestLike.scala:
```diff
@@ -0,0 +1,76 @@
+
+abstract class UnitTest extends WordSpec with UnitTestLike
+
+trait AkkaUnitTestLike extends UnitTestLike {
```
Please add some docs.

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/http/AkkaHttpResponse.scala:
```diff
@@ -0,0 +1,37 @@
+
+object AkkaHttpResponse {
+
+  def request(request: HttpRequest)(implicit
```
Please add some docs

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/mesos/FaultDomain.scala:
```diff
@@ -0,0 +1,11 @@
+package com.mesosphere.utils.mesos
```
Shouldn't this be part of some models? I don't mean `core-models` but `com.mesosphere.utils.models`.

zen-dog replied:
I'm not starting this discussion _again_. Once we get to the point when they are actually needed somewhere, we can revisit this decision.

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/mesos/MesosTest.scala:
```diff
@@ -0,0 +1,394 @@
+                        isolation: Option[String] = None,
+                        imageProviders: Option[String] = None)
+
+case class MesosCluster(
```
This definitely needs some docs.

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/mesos/MesosTest.scala:
```diff
@@ -0,0 +1,394 @@
+  def waitForLeader(): Future[String] = async {
+    val firstMaster = s"http://${masters.head.ip}:${masters.head.port}"
+    val result = Retry("wait for leader", maxAttempts = Int.MaxValue, maxDuration = waitForMesosTimeout) {
+      Http().singleRequest(Get(firstMaster + "/redirect")).map { result =>
```
What do you think about using the `MesosFacade` instead as we do for agents? We could drop `Retry` then and use `eventually`.

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/mesos/MesosTest.scala:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    def maybeFixURI(uri: String): String = {
+      // some versions of mesos issue redirects with broken Location headers; fix them here
```
Is this still the case?

zen-dog replied:
Who knows. Let's remove and see what happens

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/mesos/MesosTest.scala:
```diff
@@ -0,0 +1,394 @@
+  def waitForAgents(masterUri: String): Future[Done] = {
+    val mesosFacade = new MesosFacade(masterUri)
+    Retry.blocking("wait for agents", maxAttempts = Int.MaxValue, maxDuration = waitForMesosTimeout) {
+      val numAgents = mesosFacade.state.value.agents.size
```
See. Here we use it ;). Also, we should just use `eventually` here.

zen-dog replied:
Bam!

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/mesos/MesosTest.scala:
```diff
@@ -0,0 +1,394 @@
+    // Call mesos/teardown for all framework Ids in the cluster and wait for the teardown to complete
+    frameworkIds.foreach { fId =>
+      facade.teardown(fId)
+      eventually(timeout(1.minutes), interval(2.seconds)) { facade.completedFrameworkIds().value.contains(fId) }
```
See. We use `eventually` here.

jeschkies reviewed (changes requested):
Let's get rid off `Retry`.

zen-dog committed:
Replaced our own `Retry` implementation with scalatest `eventually`

zen-dog committed:
Added documentation

jeschkies committed:
Merge remote-tracking branch 'origin/master' into ad/test-utils

jeschkies commented:
I tried to merge master and apply `scalafmtAll` but it did not work. Any ideas why @alenkacz?

jeschkies committed:
Configure scalafmt for `test-utils`.

jeschkies committed:
Merge remote-tracking branch 'origin/master' into ad/test-utils

jeschkies committed:
Apply scalafmt for real.

zen-dog committed:
Replace `detect-routable-ip.sh` script with a Java-native method

zen-dog committed:
Fix a broken mesos-client parameter name that for some reason was broken by the scalafmt

zen-dog committed:
Fixed formatting

zen-dog committed:
Using default number of ZK connections

zen-dog committed:
Configure logging in the project

- define `logback-test.xml` for tests
- configure akka to use SLF4J in the `core-models/reference.conf` for all other project to inherit from
- moved akka-* libs to the root project and defined an `akkaVersion=2.5.19` variable for all of them
- bumped `akka-stream-alpakka-simple-codecs` to be compatible with used akka version
